libs/elasticsearch-nio/src/main/java/org/elasticsearch/nio/BytesChannelContext.java

```diff
@@ -77,19 +119,35 @@ public boolean hasQueuedWriteOps() {
     }
 
     @Override
-    public void clearQueuedWriteOps(Exception e) {
-        assert channel.getSelector().isOnCurrentThread() : "Must be on selector thread to clear queued writes";
-        for (WriteOperation op : queued) {
-            channel.getSelector().executeFailedListener(op.getListener(), e);
+    public void closeChannel() {
+        if (isClosing.compareAndSet(false, true)) {
+            channel.getSelector().queueChannelClose(channel);
+        }
+    }
+
+    @Override
+    public boolean readyToClose() {
+        return peerClosed || ioException || isClosing.get();
+    }
+
+    @Override
+    public void closeFromSelector() {
+        assert channel.getSelector().isOnCurrentThread() : "Must be on selector thread to close";
+        isClosing.set(true);
```

Comment:
so if you set this to true don't you have to close the channel as well or at least queeu it for close?

Reply:
`closeFromSelector` is only called from the channel `closeFromSelector` method (when the channel is closing. I added a comment explaining that setting this boolean to true is mainly there to reject writes in `sendMessage` to prevent enqueuing with selector. This is not critical, the selector will still reject writes that are enqueued with a closed channel. It just prevents enqueuing. 

I think this relationship is backwards. I think that the selector should call the context's `closeFromSelector` method. Which then closes the channel (similar to how the channel context controls read and writes). This is not possible in this PR as server channels do not currently have a similar concept of `ChannelContext`. I will follow up eventually to switch this relationship.

Reply:
++ thanks this makes sense to me, thanks for explaining again